src/accel/detail/HitManager.cc

```diff
@@ -168,11 +183,43 @@ void HitManager::execute(StateHostRef const& data)
  */
 void HitManager::execute(StateDeviceRef const& data)
 {
-    copy_steps(&steps_, data);
-    if (steps_)
+    auto&& process_hits = this->get_local_hit_processor();
+    process_hits(data);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Destroy local data to avoid Geant4 crashes.
+ *
+ * This deallocates the local hit processor on the Geant4 thread that was
+ * active when allocating it. This is necessary because Geant4 has thread-local
+ * allocators that crash if trying to deallocate data allocated on another
+ * thread.
+ */
+void HitManager::finalize()
+{
+    CELER_LOG_LOCAL(debug) << "Deallocating hit processor";
+    int local_thread = G4Threading::G4GetThreadId();
+    CELER_ASSERT(static_cast<std::size_t>(local_thread) < processors_.size());
+    processors_[local_thread].reset();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Ensure the local hit processor exists, and return it.
+ */
+HitProcessor& HitManager::get_local_hit_processor()
```

Comment:
All the major event loop components are "global". Only the incoming state objects are "thread-local" [^1], and normally their stream IDs should be used to access per-stream state data that (for now) has to live in the shared object. Here, instead of using G4Threading, I could (and probably should) add the `StreamId` to the `StepState` so we can access it directly...

[^1]: I put global and thread-local in quotes above because nothing in Celeritas inherently requires streams and threads to match. Streams could be shared within a thread pool, or you could have multiple streams per CPU thread, etc. Similarly you could have two entirely different event loops running simultaneously within the same application.